File: src/nodejs/src/services/OrderService.ts
import "../mappers/OrderMapper";
import { ApiCode } from "../dtos/ApiCode";
import { ApiDataResult } from "../dtos/ApiResult";
import { BaseService } from "./BaseService";
import { CreateOrderRequest } from "../dtos/CreateOrderRequest";
import { DateUtils } from "../utils/DateUtils";
import { IOrderRepository } from "../repositories/IOrderRepository";
import { IOrderService } from "./IOrderService";
import { IPatientRepository } from "../repositories/IPatientRepository";
import { Order } from "../entities/Order";
import { OrderDto } from "../dtos/OrderDto";

/**
 * Service for managing order-related business logic.
 */
export class OrderService extends BaseService implements IOrderService {
  /**
   * Creates a new instance of OrderService.
   * @param orderRepository - The order repository.
   * @param patientRepository - The patient repository.
   */
  constructor(
    private orderRepository: IOrderRepository,
    private patientRepository: IPatientRepository
  ) {
    super();
  }

  /**
   * Retrieves an order by its ID.
   * @param id - The order ID.
   * @returns An API result containing the order if found, null otherwise.
   */
  async getOrderAsync(id: number): Promise<ApiDataResult<OrderDto | null>> {
    const order = await this.orderRepository.getOrderAsync(id);

    if (!order) {
      return this.noDataFoundDataResult<OrderDto | null>(null, "Order not found");
    }

    return this.successDataResult(order.toDto());
  }

  /**
   * Creates a new order.
   * @param request - The create order request.
   * @returns An API result containing the created order.
   */
  async createOrderAsync(request: CreateOrderRequest): Promise<ApiDataResult<OrderDto>> {
    const patientExists = await this.patientRepository.isExistPatientAsync(request.patientId);

    if (!patientExists) {
      return super.failureDataResult<OrderDto>(
        ApiCode.OperationFailed,
        undefined,
        `Order with PatientId ${request.patientId} not found.`
      );
    }

    const order = new Order();
    order.patientId = request.patientId;
    order.message = request.message.trim();

    const createdOrder = await this.orderRepository.addAsync(order);
    return super.successDataResult(createdOrder.toDto());
  }

  /**
   * Updates an order's message.
   * @param id - The order ID.
   * @param message - The new message.
   * @returns An API result containing the updated order.
   */
  async updateMessageAsync(id: number, message: string): Promise<ApiDataResult<OrderDto>> {
    const updatedAt = DateUtils.utcNow();
    const updatedOrder = await this.orderRepository.updateAsync(id, message.trim(), updatedAt);

    if (!updatedOrder) {
      return super.failureDataResult<OrderDto>(ApiCode.OperationFailed, undefined, `Order with ID ${id} not found.`);
    }

    return super.successDataResult(updatedOrder.toDto());
  }
}

File: src/python/app/services/order_service.py
"""Order service."""

import logging
import redis.asyncio as redis
from datetime import datetime
from app.entities import Order
from app.repositories import OrderRepository, PatientRepository
from app.repositories.caches.cached_order_repository import CachedOrderRepository
from app.repositories.caches.cached_patient_repository import CachedPatientRepository
from app.schemas import OrderDto, CreateOrderRequest, ApiResultData, ApiCode
from sqlalchemy.ext.asyncio import AsyncSession

logger = logging.getLogger(__name__)


class OrderService:
    """Service for order business logic."""

    def __init__(self, db: AsyncSession, redis_client: redis.Redis):
        """Initialize service with database session and Redis client."""
        self.db = db
        inner_order_repo = OrderRepository(db)
        inner_patient_repo = PatientRepository(db)
        self.order_repo = CachedOrderRepository(inner_order_repo, redis_client)
        self.patient_repo = CachedPatientRepository(inner_patient_repo, redis_client)

    async def get_order(self, order_id: int) -> ApiResultData[OrderDto]:
        """
        Get order by ID.

        Args:
            order_id: Order ID

        Returns:
            API result with order data
        """
        try:
            order = await self.order_repo.get_order(order_id)

            if not order:
                return ApiResultData[OrderDto](
                    success=False, code=ApiCode.NO_DATA_FOUND, message=f"Order with ID {order_id} not found", data=None
                )

            order_dto = OrderDto.model_validate(order)
            return ApiResultData[OrderDto](
                success=True, code=ApiCode.SUCCESS, message="Order retrieved successfully", data=order_dto
            )

        except Exception as e:
            logger.error("Failed to retrieve order: %s", str(e), exc_info=True)
            return ApiResultData[OrderDto](
                success=False, code=ApiCode.DATA_ACCESS_ERROR, message=f"Failed to retrieve order: {str(e)}", data=None
            )

    async def create_order(self, request: CreateOrderRequest) -> ApiResultData[OrderDto]:
        """
        Create a new order for a patient.

        Args:
            request: Request with order data

        Returns:
            API result with created order data
        """
        try:
            patient_exists = await self.patient_repo.is_exist_patient(request.patient_id)
            if not patient_exists:
                return ApiResultData[OrderDto](
                    success=False,
                    code=ApiCode.INVALID_REQUEST,
                    message=f"Patient with ID {request.patient_id} not found",
                    data=None,
                )

            order = Order(message=request.message, patient_id=request.patient_id)

            created_order = await self.order_repo.add(order)
            await self.db.commit()
            await self.db.refresh(created_order)

            order_dto = OrderDto.model_validate(created_order)
            return ApiResultData[OrderDto](
                success=True, code=ApiCode.SUCCESS, message="Order created successfully", data=order_dto
            )

        except Exception as e:
            logger.error("Failed to create order: %s", str(e), exc_info=True)
            await self.db.rollback()
            return ApiResultData[OrderDto](
                success=False, code=ApiCode.OPERATION_FAILED, message=f"Failed to create order: {str(e)}", data=None
            )

    async def update_message(self, order_id: int, message: str) -> ApiResultData[OrderDto]:
        """
        Update order message.

        Args:
            order_id: Order ID
            message: New message

        Returns:
            API result with updated order data
        """
        try:
            utc_now = datetime.utcnow()
            updated_order = await self.order_repo.update(order_id, message.strip(), utc_now)

            if not updated_order:
                return ApiResultData[OrderDto](
                    success=False,
                    code=ApiCode.OPERATION_FAILED,
                    message=f"Order with ID {order_id} not found.",
                    data=None,
                )

            await self.db.commit()

            order_dto = OrderDto.model_validate(updated_order)

            return ApiResultData[OrderDto](
                success=True, code=ApiCode.SUCCESS, message="Order message updated successfully", data=order_dto
            )

        except Exception as e:
            logger.error("Failed to update order message: %s", str(e), exc_info=True)
            await self.db.rollback()
            return ApiResultData[OrderDto](
                success=False,
                code=ApiCode.OPERATION_FAILED,
                message=f"Failed to update order message: {str(e)}",
                data=None,
            )
